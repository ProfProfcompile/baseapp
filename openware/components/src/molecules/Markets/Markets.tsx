import * as React from 'react';
import { FilterInput } from '../';
import { CellData, Table } from '../Table/Table';

interface MarketsProps {
    /**
     * List of markets data
     */
    data: CellData[][];
    /**
     * Callback that is called when a market is selected
     */
    onSelect: (marketIndex: number) => void;
    /**
     * Defines whether to show filters or not
     * @default true
     */
    filters?: boolean;
}

interface MarketsState {
    /**
     * Keeps filtered data
     */
    filteredData: CellData[][];
}

class Markets extends React.Component<MarketsProps, MarketsState> {
    constructor(props: MarketsProps) {
        super(props);

        this.state = {
            filteredData: props.data,
        };
    }

    private headers: string[] = ['Pair', 'Price'];
    private title = 'Markets';

    public componentWillReceiveProps(nextProps: MarketsProps) {
        if (nextProps.data !== this.props.data) {
            this.setState({
                filteredData: nextProps.data,
            });
        }
    }

    public render() {
        const { filteredData } = this.state;
        const { filters = true } = this.props;
        const tableData = filteredData.length > 0
            ? filteredData
            : [['', '']];

        return (
            <div className="cr-markets">
                <Table
                    data={tableData}
                    filters={filters ? this.filters : []}
                    header={this.headers}
                    onSelect={this.props.onSelect}
                    titleComponent={this.title}
                />
                <FilterInput
                    data={this.props.data}
                    onFilter={this.handleFilter}
                    filter={this.searchFilter}
                />
            </div>
        );
    }

    public searchFilter = (row: CellData[], searchKey: string) => {
        return (row[0] as string).toLowerCase().includes(searchKey.toLowerCase());
    }

    public handleFilter = (result: object[]) => {
        this.setState({
            filteredData: [...result] as CellData[][],
        });
    }

    private filterType = (headerKey: string, searchKey: string) => (item: CellData[]) => {
        const typeIndex = this.headers.indexOf(headerKey);
        return (item[typeIndex] as string).includes(searchKey);
    };

    private get filters() {
        const { data } = this.props;

        const currencyFilters = data && data.length > 0
            ? this.props.data
                .map(this.getMarketFromDataRow)
                .reduce(this.createUniqueCurrencies, [])
                .map(this.transformCurrencyToFilter)
            : [];

        return [
            {
                name: 'All',
                filter: this.filterType('Pair', ''),
            },
            ...currencyFilters,
        ];
    }

    private getMarketFromDataRow = (market: React.ReactNode[]) => market[0] as string;

    private createUniqueCurrencies(currencies: string[], market: string) {
        const isCurrencyUnique = (currency: string) => !currencies.includes(currency);

        const marketCurrencies = market.split('/').map((c: string) => c.trim());
        const uniqueCurrencies = marketCurrencies.filter(isCurrencyUnique);

        return currencies.concat(uniqueCurrencies);
    }

    private transformCurrencyToFilter = (currency: string) => ({
        name: currency,
        filter: this.filterType('Pair', currency),
    });
}

export {
    Markets,
    MarketsProps,
    MarketsState,
};
