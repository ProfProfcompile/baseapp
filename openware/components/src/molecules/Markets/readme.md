```js
const data = [
  ['ETH / BTC', '0.0710000'],
  ['XPR / BTC', '0.0710000'],
  ['XPR / LTC', '0.0710000'],
  ['USDT / BTC', '0.0710000'],
  ['LTC / BTC', '0.0710000'],
];

const onSelect = (index) => console.log(index);

<Markets data={data} onSelect={onSelect} />
```