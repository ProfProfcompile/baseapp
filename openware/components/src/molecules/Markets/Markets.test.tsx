import { mount, ReactWrapper, shallow, ShallowWrapper } from 'enzyme';
import * as React from 'react';
import { spy } from 'sinon';
import { CellData } from '../Table/Table';
import { Markets, MarketsProps, MarketsState } from './Markets';

const data: CellData[][] = [
    ['ETH / BTC', '0.223100'],
    ['ETH / LTC', '0.223100'],
    ['LTC / BTC', '0.223100'],
];
const onSelect = spy();

const defaultProps: MarketsProps = {
    data,
    onSelect,
};

const setup = (props?: Partial<MarketsProps>) =>
    shallow(<Markets {...{ ...defaultProps, ...props }} />);

describe('Markets', () => {
   let wrapper: ShallowWrapper;

   beforeEach(() => {
      wrapper = setup();
   });

   it('should render', () => {
       expect(wrapper).toMatchSnapshot();
   });

   it('should render empty data', () => {
       wrapper = setup({ data: [] });
       expect(wrapper).toMatchSnapshot();
   });

   it('should correctly filter rows', () => {
       expect((wrapper.instance() as Markets).searchFilter(['ETH / BTC', '0.123'], 'btc')).toBeTruthy();
       expect((wrapper.instance() as Markets).searchFilter(['ETH / BTC', '0.342'], 'ltc')).toBeFalsy();
   });

   it('should set filtered data to state', () => {
      const component: ReactWrapper = mount(
          <Markets data={data} onSelect={onSelect} />,
      );

      const filteredData: CellData = [
          ['ETH / BTC', '0.123'],
      ];

      (component.instance() as Markets).handleFilter(filteredData as object[]);

      expect((component.state() as MarketsState).filteredData).toEqual(filteredData);
   });

   it('should set new data to state', () => {
       const component: ReactWrapper = mount(
           <Markets data={data} onSelect={onSelect} />,
       );

       const filteredData: CellData = [
           ['ETH / BTC', '0.123'],
       ];

       component.setProps({ data: filteredData });
       expect((component.state() as MarketsState).filteredData).toEqual(filteredData);
   });
});
