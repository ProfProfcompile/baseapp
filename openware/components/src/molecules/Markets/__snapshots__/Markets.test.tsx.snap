// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Markets should render 1`] = `
<div
  className="cr-markets"
>
  <Table
    data={
      Array [
        Array [
          "ETH / BTC",
          "0.223100",
        ],
        Array [
          "ETH / LTC",
          "0.223100",
        ],
        Array [
          "LTC / BTC",
          "0.223100",
        ],
      ]
    }
    filters={
      Array [
        Object {
          "filter": [Function],
          "name": "All",
        },
        Object {
          "filter": [Function],
          "name": "ETH",
        },
        Object {
          "filter": [Function],
          "name": "BTC",
        },
        Object {
          "filter": [Function],
          "name": "LTC",
        },
      ]
    }
    header={
      Array [
        "Pair",
        "Price",
      ]
    }
    onSelect={[Function]}
    titleComponent="Markets"
  />
  <FilterInput
    data={
      Array [
        Array [
          "ETH / BTC",
          "0.223100",
        ],
        Array [
          "ETH / LTC",
          "0.223100",
        ],
        Array [
          "LTC / BTC",
          "0.223100",
        ],
      ]
    }
    filter={[Function]}
    onFilter={[Function]}
  />
</div>
`;

exports[`Markets should render empty data 1`] = `
<div
  className="cr-markets"
>
  <Table
    data={
      Array [
        Array [
          "",
          "",
        ],
      ]
    }
    filters={
      Array [
        Object {
          "filter": [Function],
          "name": "All",
        },
      ]
    }
    header={
      Array [
        "Pair",
        "Price",
      ]
    }
    onSelect={[Function]}
    titleComponent="Markets"
  />
  <FilterInput
    data={Array []}
    filter={[Function]}
    onFilter={[Function]}
  />
</div>
`;
